src/tests/java/picard/illumina/ExtractIlluminaBarcodesTest.java

```diff
@@ -105,18 +105,35 @@ public void testSingleEndWithBarcodeAtStart() throws Exception {
     }
 
     @Test
+    public void testSingleEndWithBarcodeAtStartAndMolecularIndicies() throws Exception {
+        final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "8B4M21T");
+        Assert.assertEquals(metricsFile.getMetrics().get(11).PERFECT_MATCHES, 1);
+    }
+
+    @Test
     public void testSingleEndWithBarcodeAtEnd() throws Exception {
         final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "25T8B");
         Assert.assertEquals(metricsFile.getMetrics().get(0).PERFECT_MATCHES, 5);
     }
 
     @Test
+     public void testSingleEndWithBarcodeAtEndAndMolecularIndicies() throws Exception {
+        final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "4M21T8B");
+        Assert.assertEquals(metricsFile.getMetrics().get(0).PERFECT_MATCHES, 5);
+    }
+
+    @Test
     public void testPairedEndWithBarcodeOnFirstEnd() throws Exception {
         final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "25T8B25T");
         Assert.assertEquals(metricsFile.getMetrics().get(0).PERFECT_MATCHES, 5);
     }
 
     @Test
+    public void testPairedEndWithBarcodeOnFirstEndAndMolecularIndicies() throws Exception {
+        final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "4M21T8B21T4M");
+        Assert.assertEquals(metricsFile.getMetrics().get(0).PERFECT_MATCHES, 5);
+    }
+    @Test
     public void testPairedEndWithBarcodeOnSecondEnd() throws Exception {
         final MetricsFile<ExtractIlluminaBarcodes.BarcodeMetric, Integer> metricsFile = runIt(1, "25T25T8B");
         Assert.assertEquals(metricsFile.getMetrics().get(12).PERFECT_MATCHES, 1);
```

Comment:
While trying to familiarize myself with `ReadStructure.java`, does this mean that `4M21T8B21T4M` would be:
- read one 4 cycles of molecular index
- read two 21 cycles of template
- read three 8 cycles of sample barcode
- read four 21 cycles of template
- read five 4 cycles of molecular index
?

Perhaps also updating the docs and providing an example in `ReadStructure.java` would be appropriate.